airbyte-workers/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java

```diff
@@ -116,16 +122,53 @@ private void handleSourceEmittedRecord(final AirbyteRecordMessage recordMessage)
     streamToTotalBytesEmitted.put(streamIndex, currentTotalStreamBytes + estimatedNumBytes);
   }
 
+  /**
+   * Save the lastest state on the provided output state. It is using the {@link StateLifecycleManager} to manage the state transition.
+   * The markState method is a runnable to call the right method to mark a state and the state supplier provide the states that are updated teh last.
+   *
+   * @param stateMessage - the state message to ingest
+   * @param outputState - atomic reference on the state to update (source or destination)
+   * @param markState - runnnable whose responsability is to call the right mark method
+   * @param stateSupplier - Supplier to get the state with the right formats
+   * @return
+   */
+  private AirbyteStateMessage saveStateOutput(final AirbyteMessage stateMessage,
+                                              final AtomicReference<State> outputState,
+                                              final Runnable markState,
+                                              final Supplier<Queue<AirbyteMessage>> stateSupplier) {
+    stateLifecycleManager.addState(stateMessage);
+    markState.run();
+    final Queue<AirbyteMessage> states = stateSupplier.get();
+    AirbyteStateMessage latestState = stateMessage.getState();
+    while (!states.isEmpty()) {
+      final AirbyteStateMessage airbyteStateMessage = states.poll().getState();
+      latestState = airbyteStateMessage;
+      if (airbyteStateMessage.getStateType() == null) {
+        outputState.set(new State().withState(airbyteStateMessage.getData()));
+      } else {
+        switch (airbyteStateMessage.getStateType()) {
+          case GLOBAL, STREAM -> outputState.set(new State().withState(Jsons.jsonNode(Lists.newArrayList(airbyteStateMessage))));
+          case LEGACY -> outputState.set(new State().withState(airbyteStateMessage.getData()));
+        }
+      }
+      totalEmittedStateMessages.incrementAndGet();
+    }
+
+    return latestState;
+  }
+
   /**
    * When a source emits a state, persist the current running count per stream to the
    * {@link StateDeltaTracker}. Then, reset the running count per stream so that new counts can start
    * recording for the next state. Also add the state to list so that state order is tracked
    * correctly.
    */
-  private void handleSourceEmittedState(final AirbyteStateMessage stateMessage) {
-    sourceOutputState.set(new State().withState(stateMessage.getData()));
-    totalEmittedStateMessages.incrementAndGet();
-    final int stateHash = getStateHashCode(stateMessage);
+  private void handleSourceEmittedState(final AirbyteMessage airbyteMessage) {
+    final AirbyteStateMessage latestState = saveStateOutput(airbyteMessage,
+        sourceOutputState,
+        () -> stateLifecycleManager.markPendingAsFlushed(),
+        () -> stateLifecycleManager.listFlushed());
```

Comment:
I'm really glad you put up this PR early. I think that reusing the lifecycle manager is actually not going to work.

I think this isn't quite right. One of the the drawbacks about reusing this state lifecycle manager here is that the concept of `flushed` doesn't really exist in the worker; it only makes sense in the destination.

basically the worker knows 2 concepts:
1. `pending` the state message was emitted by the source, but we haven't gotten it back from the destination yet
2. `committed` we got the state message from the destination

i think i'm realizing that reusing this lifecycle manager is maybe a mistake. the lifecycle is actually very specific to the destination and makes some assumptions. 

destination assumptions:
* at a given moment in time all records in pending are moved from pending to flushed (this may happen multiple times, but each time it does, all records that are pending at that moment move to flushed)
* at a given moment in time all records in flushed are moved to committed

worker assumptions:
* a record emitted from the source is "pending"
* a record emitted from the destination is "committed"
* no guarantee that when a single record is emitted from the destination that all pending records are committed. only that record is committed.

Reply:
I agree with you. we are not really using the `pending` status from the lifecycle manager we directly go to `flush` and then `complete`. 
For the third point of the worker assumption, I understand that the tracker will be call like that:
```
acceptFromSource(m1)
acceptFromSource(m2)
acceptFromDestination(m1)
acceptFromSource(m3)
acceptFromDestination(m2)
acceptFromDestination(m3)
```
With every message being consider as pending or committed independently. Is that what you mean?